viamonvifdiscovery/service.go

```diff
@@ -0,0 +1,201 @@
+			}
+		}
+	}
+}
+
+// Close closes the discovery service.
+func (dis *rtspDiscovery) Close(_ context.Context) error {
+	return nil
+}
+
+// DiscoverResources discovers different rtsp cameras that use onvif.
+func (dis *rtspDiscovery) DiscoverResources(_ context.Context, _ map[string]any) ([]resource.Config, error) {
+	potentialCams := []resource.Config{}
+	insecureURLs := []string{}
+
+	for _, cred := range dis.Credentials {
+		// if we have multiple empty/insecure credentials configured for some reason, skip to the next credential.
+		// only skip if we have not processed insecure credentials yet.
+		if cred.isInsecure() && len(insecureURLs) > 0 {
+			continue
+		}
+
+		list, err := viamonvif.DiscoverCameras(cred.Username, cred.Password, dis.logger, nil)
+		if err != nil {
+			return nil, err
+		}
+		cameraNumber := 0
+		for _, l := range list.Cameras {
+			dis.logger.Debugf("%s %s %s", l.Manufacturer, l.Model, l.SerialNumber)
+			// some cameras return with no urls. explicitly skipping those so the behavior is clear in the service.
+			if len(l.RTSPURLs) == 0 {
+				dis.logger.Debugf("No urls found for camera, skipping. %s %s %s", l.Manufacturer, l.Model, l.SerialNumber)
+				continue
+			}
+			cameraNumber++
+			camConfigs, err := createCamerasFromURLs(cred.createName(cameraNumber), l, insecureURLs, dis.logger)
+			if err != nil {
+				return nil, err
+			}
+			potentialCams = append(potentialCams, camConfigs...)
+
+			// if we are processing insecure creds, record the urls so we do not duplicate the cameras when processing cameras with creds.
+			if cred.isInsecure() {
+				insecureURLs = append(insecureURLs, l.NoLoginURLs...)
+			}
+		}
+	}
+	return potentialCams, nil
+}
+
+func createCamerasFromURLs(cameraName string, l viamonvif.CameraInfo, insecureCams []string,
+	logger logging.Logger,
+) ([]resource.Config, error) {
+	potentialCams := []resource.Config{}
+	for index, u := range l.RTSPURLs {
+		// skip over cameras that were already discovered with insecure creds
+		if slices.Contains(insecureCams, l.NoLoginURLs[index]) {
+			continue
+		}
+		logger.Debugf("\t%s", u)
+		cfg, err := createCameraConfig(cameraName, u)
+		if err != nil {
+			return nil, err
+		}
+		potentialCams = append(potentialCams, cfg)
+	}
+	return potentialCams, nil
+}
+
+// set the camera name based on the Username and camera number.
+func (cred *Creds) createName(index int) string {
+	if cred.Username == "" {
+		return fmt.Sprintf("Camera_Insecure_%v", index)
+	}
+	if cred.credNumber > 0 {
+		return fmt.Sprintf("Camera_%s-%v_%v", cred.Username, cred.credNumber, index)
+	}
+	return fmt.Sprintf("Camera_%s_%v", cred.Username, index)
+}
+
+func (cred *Creds) isInsecure() bool {
+	return cred.Username == "" && cred.Password == ""
+}
+
+func createCameraConfig(name, address string) (resource.Config, error) {
+	// using the camera's Config struct in case a breaking change occurs
+	attributes := viamrtsp.Config{Address: address}
+	var result map[string]interface{}
```

Comment:
Is this used anywhere? I would expect that we could populate the attributes - I'm uncertain if we have to without a front end. But something to note for in case flutter needs this part of the robot config to populate the mobile app config - probably important context for martha.

Reply:
The result is part of the returned config, unless I’m misunderstanding what “this” is here.